CPLAT-7558 Add Component2 Documentation (#377)
Workiva/over_react: doc/ui_component2_transition.md

> A document to explain the new features and breakages when converting components from Component / UiComponent to 
Component2 / UiComponent2.
---
* __[What's New in Component2](#whats-new-in-component2)__
* __[Why a new base class?](#why-a-new-base-class)__
* __[New Features, improvements, and breakages](#new-features-improvements-and-breakages)__
    * [Small Perf Improvements](#small-perf-improvements)
    * [JS-Backed Maps](#js-backed-maps)
    * [Lifecycle Method Changes](#lifecycle-method-changes)
    * [Other Method Changes](#other-method-changes)
    * [PropTypes](#prop-validation-goes-through-proptypes)
    * [ErrorBoundary Component Added](#errorboundary-component-added)
    * [Other Breaking Changes](#other-breakages-we-made-while-we-had-the-opportunity)
* __[Updating](#updating)__
## What's New in Component2
`Component` and `UiComponent` are being deprecated in favor of new versions of the API: `Component2` and `UiComponent2`.

#### tl;dr

- Pros: slightly faster, improved dev experience, easier to maintain, easier integration with JS libs
- Cons: breaks a few advanced component APIs that will require conversion, some of which will be automated with codemod

> Migration can happen at any time, and is opt-in. We'll provide a [codemod](#updating) to handle the simpler cases and 

Reviewer: ```suggestion
Migration to UiComponent2 is opt-in and can happen at any time until the next major release of OverReact. A [codemod](#updating) is available to handle the simpler cases and we will
```

run it as 
part of cleanup, but it will take manual intervention to update some components.

## Why a new base class?
- It's not possible for us to automatically support both unsafe and safe lifecycle methods without a resolved AST in 
the over_react builder, which would increase build times dramatically.
- We needed some small breakages to support
  - JS-backed maps
  - New lifecycle methods
  - Injection of Dart->JS interop pieces into component instance (initComponentInternal) needed to change (this was 

>  publicly accessible, but is not done anywhere)

Reviewer: done or used?

Author: Hmm pretty sure used. Good catch!

  - componentDidUpdate, which now takes an additional `snapshot` argument

>   - Assignment of `props`/`state` to arbitrary Maps (was not advised, but possible)

Reviewer: This line is confusing to me, did we make it possible and it IS not advised or was it possible (and not advised) and it is not possible anymore?

Author: I'm 99% sure we made it impossible to set the maps directly, and I updated the wording to reflect that, but leaving this unresolved for @greglittlefield-wf to confirm

Reviewer 2: Now they can only be set to `JsBackedMap`s, but it's considered unsupported usage.

- Modest overall rendering perf improvements due to decreased copying, garbage creation, and Dart map overhead
- Props/state show up in the React DevTools, and primitive values (strings, numbers, booleans) can be edited live
- Easier to interop with JS libraries that manipulate props, like react-redux
- Easier to maintain react-dart/over_react interop and add new features

### Lifecycle Method Changes

- Drops support for unsafe lifecycle methods:
  - componentWillMount
  - componentWillReceiveProps
  - componentWillUpdate
- Adds support for new, safe lifecycle methods:
  - getDerivedStateFromProps
  - getSnapshotBeforeUpdate
  - componentDidCatch / getDerivedStateFromError (adds support for [error boundaries](https://reactjs.org/docs/error-boundaries.html))
- componentDidUpdate now receives a `snapshot` argument, which is the return value of getSnapshotBeforeUpdate.
  
See the [React docs](https://reactjs.org/docs/react-component.html#the-component-lifecycle) for information on why the old methods are unsafe and how to transition to the new ones. This 
information is also available and linked to within the lifecycle methods' doc comments.

> ### ErrorBoundary Component Added

Reviewer: I feel like this section should go after "Other Method Changes" or before "Lifecycle Method Changes" for continuity.

Author: Ah yep, good call 😄

With the ability to utilize componentDidCatch / getDerivedStateFromError comes the ability to use [error boundaries](https://reactjs.org/docs/error-boundaries.html).
A component (called ErrorBoundary) is present within OverReact and can be used to wrap a component tree to prevent it 
from unmounting when an error is thrown. Alternatively, a custom error boundary can be created by utilizing the 
[ErrorBoundaryMixin](../lib/src/component/error_boundary_mixins.dart).

Usage of the default error boundary can be found in the Component1 examples (within web), and usage of a custom error 
boundary can be found in the Component2 examples.

### Other Method Changes
- `getDefaultProps` is now `get defaultProps`
- `getInitialState` is now `get initialState`
- `..addProps(copyUnconsumedProps())` is now `..modifyProps(addUnconsumedProps)`. The same is true with `DomProps`, 

> the usage being `..modifyProps(addUnconsumedDomProps)`.

Reviewer: is it worth mentioning the perf improvement that this has compared to the previous way?

Author: I believe so!

> ### Prop validation goes through PropTypes

Reviewer: ```suggestion
### Prop validation via PropTypes
```

- Shows "component stack" that helps determine where the error came from
- React automatically de-duplicates the same warnings

> - __No longer throws, but warns__. If tests exist verifying that prop validation occurs, these will likely need to 

Reviewer: We should mention that this only applies if they are update the component to UiComponent2

change. There are new `PropType` testing utils within [over_react_test](https://github.com/Workiva/over_react_test).

> 

Reviewer: ```suggestion
- Aligns with ReactJS implementation

```

# UiComponent2 Transition Notes
> A document to explain the new features and breakages when converting components from Component / UiComponent to 
Component2 / UiComponent2.
---
* __[What's New in Component2](#whats-new-in-component2)__
* __[Why a new base class?](#why-a-new-base-class)__

> * __[New Features, improvements, and breakages](#new-features-improvements-and-breakages)__

Reviewer 2: #nit inconsistent casing
```suggestion
* __[New Features, Iimprovements, and Breakages](#new-features-improvements-and-breakages)__
```

    * [Small Perf Improvements](#small-perf-improvements)
    * [JS-Backed Maps](#js-backed-maps)
    * [Lifecycle Method Changes](#lifecycle-method-changes)
    * [Other Method Changes](#other-method-changes)
    * [PropTypes](#prop-validation-goes-through-proptypes)
    * [ErrorBoundary Component Added](#errorboundary-component-added)
    * [Other Breaking Changes](#other-breakages-we-made-while-we-had-the-opportunity)
* __[Updating](#updating)__

> ## What's New in Component2

Reviewer 2: I think having this as a separate section, at least with this title, isn't super helpful. This is really an intro or overview section.

Perhaps we can just rename this?
```suggestion
## Overview
```

> `Component` and `UiComponent` are being deprecated in favor of new versions of the API: `Component2` and `UiComponent2`.

Reviewer 2: Since this document lives in the repo, I think it'd be helpful to add a little bit of context as to when this occurred.

Something like:
```suggestion
In `react` 5.1.0 and `over_react` 3.1.0, `Component` and `UiComponent` were deprecated in favor of new versions of the API: `Component2` and `UiComponent2`.
```

- Cons: breaks a few advanced component APIs that will require conversion, some of which will be automated with codemod

Migration to UiComponent2 is opt-in and can happen at any time until the next major release of OverReact. A 
[codemod](#updating) is available to handle the simpler cases, but it will take manual intervention to update some 
components.

## Why a new base class?
- It's not possible for us to automatically support both unsafe and safe lifecycle methods without a resolved AST in 
the over_react builder, which would increase build times dramatically.
- We needed some small breakages to support
  - JS-backed maps
  - New lifecycle methods
  - Injection of Dart->JS interop pieces into component instance (initComponentInternal) needed to change (this was 
 publicly accessible, but was not used anywhere)
  - componentDidUpdate, which now takes an additional `snapshot` argument
  - Make assignment of `props`/`state` to arbitrary Maps impossible (was not advised, but possible)
  - Prop validation, which had to be reimplemented since it relied on unsafe component lifecycle transitions

> ## New Features, improvements, and breakages

Reviewer 2: #nit inconsistent casing
```suggestion
## New Features, Improvements, and Breakages
```

### Small Perf Improvements
Component2 provides a small performance boost over Component, as a result of JS-backed maps and other optimizations.

> #### Data

Reviewer 2: Can we add in the graphs to this section? I think those are pretty helpful visual aides, and also show the difference between Dart and pure JS components

### Lifecycle Method Changes

- Drops support for unsafe lifecycle methods:
  - componentWillMount
  - componentWillReceiveProps
  - componentWillUpdate
- Adds support for new, safe lifecycle methods:
  - getDerivedStateFromProps
  - getSnapshotBeforeUpdate
  - componentDidCatch / getDerivedStateFromError (adds support for [error boundaries](https://reactjs.org/docs/error-boundaries.html))
- componentDidUpdate now receives a `snapshot` argument, which is the return value of getSnapshotBeforeUpdate.
  
See the [React docs](https://reactjs.org/docs/react-component.html#the-component-lifecycle) for information on why the old methods are unsafe and how to transition to the new ones. This 
information is also available and linked to within the lifecycle methods' doc comments.

### Other Method Changes
- `getDefaultProps` is now `get defaultProps`
- `getInitialState` is now `get initialState`
- `..addProps(copyUnconsumedProps())` is now `..modifyProps(addUnconsumedProps)`. The same is true with `DomProps`, 
the usage being `..modifyProps(addUnconsumedDomProps)`. Using `addUnconsumedProps` was shown to render 8% faster than

>  `copyUncosumedProps`.

Reviewer 2: typo
```suggestion
 `copyUnconsumedProps`.
```

### Prop validation via PropTypes

> - Aligns with ReactJS implementation.

Reviewer 2: #nit the other short bullets don't have periods
```suggestion
- Aligns with ReactJS implementation
```

### ErrorBoundary Component Added

With the ability to utilize componentDidCatch / getDerivedStateFromError comes the ability to use [error boundaries](https://reactjs.org/docs/error-boundaries.html).
A component (called ErrorBoundary) is present within OverReact and can be used to wrap a component tree to prevent it 
from unmounting when an error is thrown. Alternatively, a custom error boundary can be created by utilizing the 
[ErrorBoundaryMixin](../lib/src/component/error_boundary_mixins.dart).

Usage of the default error boundary can be found in the Component1 examples (within web), and usage of a custom error 
boundary can be found in the Component2 examples.

### Other breakages we made while we had the opportunity
- Split `Component.setState` into two methods, so that each has strong typing 
  - setState's first argument is now typed as `Map`, not `dynamic`
  - A new `setStateWithUpdater` is added, which accepts transactional updater functions
- Remove `bind`—a non-standard API leftover from early react-dart code—to reduce perf overhead
- Re-implement prop validation and remove support for `validateProps`
- Drop support for legacy context APIs

## Updating

> To update your repository to use UiComponent2 / Component2, you can use 

Reviewer 2: I think it'd be worth making it clear up front that the codemod won't automatically upgrade everything perfectly.

> [over_react_codemod](https://github.com/Workiva/over_react_codemod)'s component2_upgrade executable. This codemod 

Reviewer 2: #nit code formatting for this executable might be nice to have it stand out
```suggestion
[over_react_codemod](https://github.com/Workiva/over_react_codemod)'s `component2_upgrade` executable. This codemod 
```
